src/operator/nn/dnnl/dnnl_reduce.cc

```diff
@@ -0,0 +1,236 @@
+
+namespace mxnet {
+namespace op {
+
+template <>
+NumpyReduceAxesParam ConvertParamsToNumpy<ReduceAxesParam>(const ReduceAxesParam& original_param,
+                                                           const NDArray& input,
+                                                           const NDArray& output) {
+  NumpyReduceAxesParam numpy_param;
+  numpy_param.axis = dmlc::optional<mxnet::Tuple<int>>();
+  if (original_param.axis.has_value()) {
+    mxnet::Tuple<int> axes(original_param.axis.value().begin(), original_param.axis.value().end());
+    std::sort(axes.begin(), axes.end());
+
+    if (original_param.exclude) {
+      const size_t in_ndim = input.shape().ndim();
+      mxnet::Tuple<int> inverted_axes(in_ndim - axes.ndim(), -1);
+      for (int i = 0, j = 0; i < input.shape().ndim(); i++) {
+        if (j >= axes.ndim() || i != axes[j]) {
+          inverted_axes[i - j] = i;
+        } else {
+          j++;
+        }
+      }
+      numpy_param.axis = inverted_axes;
+    } else {
+      numpy_param.axis = axes;
+    }
+  }
+  numpy_param.keepdims = original_param.keepdims;
+  numpy_param.dtype    = dmlc::optional<int>(output.dtype());
+  return numpy_param;
+}
+
+template <>
+NumpyReduceAxesParam ConvertParamsToNumpy<NumpyReduceAxesParam>(
+    const NumpyReduceAxesParam& original_param,
+    const NDArray& input,
+    const NDArray& output) {
+  return original_param;
+}
+
+mxnet::Tuple<int> CanonicalizeAndSortAxes(const NDArray& input,
+                                          const NumpyReduceAxesParam& param,
+                                          mxnet::Tuple<int> original_axes) {
+  int in_ndim = input.shape().ndim();
+  mxnet::Tuple<int> axes(param.axis.value());
+  // canonicalize
+  for (index_t i = 0; i < axes.ndim(); i++) {
+    if (axes[i] < 0) {
+      axes[i] += in_ndim;
+    }
+  }
+  std::sort(axes.begin(), axes.end());
+  return axes;
+}
+
+bool SupportDNNLReduceImpl(const NumpyReduceAxesParam& param,
+                           const NDArray& input,
+                           const NDArray& output) {
+  int in_ndim          = input.shape().ndim();
+  int out_size         = output.shape().Size();
+  int in_size          = input.shape().Size();
+  bool param_supported = true;
+  if (param.axis.has_value()) {
+    auto axes    = CanonicalizeAndSortAxes(input, param, param.axis.value());
+    int last_dim = *(axes.end() - 1);
+    if (last_dim != input.shape().ndim() - 1) {
+      // oneDNN not optimized case
+      return false;
+    } else {
+      for (int i = 0; i < axes.ndim(); i++) {
+        // oneDNN doesnt support reduction of axes with dimension 1
+        // use oneDNN implementation only when dealing with consecutive trailing dimensions
+        if (input.shape()[axes[i]] == 1 || (last_dim - axes[i]) != (axes.ndim() - 1 - i)) {
+          return false;
+        }
+      }
+    }
+
+    // if axis = () it's identity op and it is not supported by oneDNN
```

Comment:
Maybe add `` around `axis = () for readability? `it's` -> `it is` to be consistent with `it is not`.

Reply:
done